plugins/inputs/t128_tank/reader.go

```diff
@@ -0,0 +1,354 @@
+	}()
+
+	go r.read(mainCtx, readCtx, lastIndex.next())
+
+	for {
+		select {
+		case <-mainCtx.Done():
+			fmt.Printf("%s reader done", r.topic)
+			readCtxCancel()
+			return
+		case <-nextSaveCheck:
+			if lastIndex.value > r.lastSavedIndex {
+				setIndex(r.indexPath, lastIndex)
+				r.lastSavedIndex = lastIndex.value
+			}
+			nextSaveCheck = time.After(2 * time.Second)
+		case err := <-r.readDone:
+			if !paused {
+				readCtx, readCtxCancel = context.WithCancel(mainCtx)
+				var boundaryFault *boundaryFault
+				if err != nil && errors.As(err, &boundaryFault) {
+					fmt.Printf("detected boundary fault, restarting %s tank read from index %d", r.topic, boundaryFault.nextAvailableIndex.value)
+					go r.read(mainCtx, readCtx, boundaryFault.nextAvailableIndex)
+				} else {
+					go r.read(mainCtx, readCtx, lastIndex.next())
+				}
+			} else {
+				fmt.Printf("reader is paused, skipping %s tank read restart", r.topic)
+			}
+		}
+	}
+}
+
+func (r *Reader) read(mainCtx context.Context, readCtx context.Context, startingIndex index) {
+	fmt.Printf("starting %s tank read from index %d", r.topic, startingIndex.value)
+	err := r.readFromTank(readCtx, startingIndex)
+	if err != nil {
+		fmt.Printf("%s read routine exited with error: %v", r.topic, err)
+	}
+
+	if mainCtx.Err() == nil {
+		time.Sleep(r.restartDelay)
+	}
+
+	r.readDone <- err
+}
+
+func (i *IndexedMessage) WithTimestamp(timestamp int) []byte {
+	parts := strings.Fields(string(i.Message))
+	parts[len(parts)-1] = strconv.Itoa(timestamp)
+	return []byte(strings.Join(parts, " "))
+}
+
+func (i *IndexedMessage) IsValid() bool {
+	return len(i.Message) > 0
+}
+
+func parseLines(tankReader *bufio.Reader, topic string) (messages []*IndexedMessage, err error) {
+	messages = make([]*IndexedMessage, 0, 1)
+	rawLines, err := tankReader.ReadBytes('\n')
+	if err != nil && !errors.Is(err, io.EOF) {
+		return messages, err
+	} else if len(rawLines) == 0 {
+		return messages, fmt.Errorf("found empty newline in %s tank stream, continuing", topic)
+	}
+	rawLines = bytes.TrimSpace(rawLines)
+	lines := bytes.Split(rawLines, []byte("\n"))
+	for _, rawLine := range lines {
+		line := string(rawLine)
+
+		if len(line) == 0 {
+			continue
+		}
+		message, err := ParseTankLine(rawLine)
+		if err != nil {
+			fmt.Printf("parse failure for topic %s: %v", topic, err)
+			continue
+		}
+
+		if message == nil {
+			if isBoundaryFault, nextAvailabelIndex := isBoundaryFault(line); isBoundaryFault {
+				return messages, fmt.Errorf(
+					"parsed boundary fault: %w",
+					&boundaryFault{nextAvailableIndex: nextAvailabelIndex},
+				)
+			}
+
+			fmt.Printf("skipping %s line because regex matching failed: %s", topic, line)
+			continue
+		}
+
+		message.Message = bytes.TrimRight(message.Message, "\n")
+		messages = append(messages, message)
+	}
+
+	return messages, err
+}
+
+func ParseTankLine(line []byte) (*IndexedMessage, error) {
+	matches := indexRegex.FindSubmatch(line)
```

Comment:
We do not want to execute a regex on each line. There's an update with a better way I thought I gave you.
- [ ] Regex on each line is way too slow